### DISCUSSION TITLE
The ability to rename columns in table layout, especially for those that are properties of relations

### DISCUSSION THREAD

**Discussion by sxsws** [directus/directus#13277] [Feature Requests]
It'd be nice to be able to rename headers in index views, because some relations can have similar sounding fields that make sense in the context of that object, but not via the relation, e.g. `school` has `name` and may have a relation to a `county` and `county.name` will display as `name` in the index view.

Being able to rename it to e.g. `County Name` would help improve usability.

I suppose a quick fix would be to automatically title headers that are for a relation to, e.g. `County.Name` rather than just `Name`.

**Comment by tenebrius** [comment]
Any progress on this idea?

**Comment by rasco** [comment]
I think the quick fix would be easy to implement and would be VERY helpful. 
I'd propose the format `Collection Name` (with a space instead of with a dot as above).

**Reply by sxsws** [reply]
This wouldn't work for more than one relation deep though, e.g. a collection of `Persons` and want to show their postcode which is a collection, so `address.postcode.code` would become `Address Postcode Code`.

**Comment by sxsws** [comment]
If someone wants to poke around I'm going to take a guess that it could be in `useTable` function in `directus/app/src/layouts/tabular/index.ts`

I've also just found out that if you hover over the dot in the column header of relation fields, you see the 'tree' of the relation like `relationFieldName -> somePropertyName`. This string is built in the `description` part of the `useTable` method, so...  this seems like a really, REALLY, simple PR infact!

I'd imagine something like this should do the trick! (-/+ indicating the only two changes necessary). Code not checked!

```
if (fieldParts.length > 1) {
	const fieldNames = fieldParts.map((fieldKey, index) => {
		const pathPrefix = fieldParts.slice(0, index);
		const field = fieldsStore.getField(collection.value!, [...pathPrefix, fieldKey].join('.'));
		return field?.name ?? fieldKey;
	});

	description = fieldNames.join(' -> ');
+      relationalFieldName = fieldNames.join('.');
}

return {
-       text: field.name
+	text: fieldParts.length > 1 ? relationalFieldName : field.name,
```

I love the " -> " annotation between collections and properties for the description, but I think in terms of horizontal space needed for headers, it may consume a bit too much.

Also, semi-related: https://github.com/directus/directus/discussions/13614 - the optional field description could be shown in the field header too.

**Reply by tenebrius** [reply]
I think this is a very reasonable solution. Should we submit a PR?

**Reply by sxsws** [reply]
I'll do shortly. I have a few other ideas that I may have a play with.

**Reply by rijkvanzanten** [reply]
FWIW I'd prefer to keep the `->` style (as with the Inter font, that becomes a neat ligature) over the `.` syntax. Just keeps it a little more human readable / less technical 🙂

**Reply by sxsws** [reply]
I agree it is nice and friendlier but I'm just concerned about the horizontal space it would take up over a `.`.

**Reply by rijkvanzanten** [reply]
Yup! Which is the reason why it's not shown currently (but instead shows the `⚪` that you can hover on for the full path). Another problem is that by putting the field name last, it's the first thing that gets cut off on smaller column sizes, so there's a good chance you'd end up with `Autho... | Autho... | Autho...` instead of `⚪ Name | ⚪ Age | ⚪ Locat...`

**Reply by sxsws** [reply]
Exactly right, and I was thinking about using the quick idea as a quickfix until column header renaming is possible. 

My original idea was to be able to click on the header, have a modal pop up and enter the new column header name (along with translations if necessary) but  just earlier today thought about moving it to the side menu, where in the options section for the tabular view, could be some sort of list that represents the visible columns, with a title input (with translations) that would allow entering a new string for the column header.

**Reply by rijkvanzanten** [reply]
Where would those custom names be saved? I suppose it would be on the preset of the layout options, though those are per collection and per role/user, so maybe that's not super in the UX dept. You wouldn't want to use the data-model settings for this either, as you do in fact want to scope it per layout I suppose, as it's only really mostly relevant for fields you nest through a relationship, however, if it's per layout that config isn't shared/reused between different layouts, which would be pretty inconvenient if you switch from like a configured map preview to the tabular layout 🤔

**Reply by AustinPhillipTaylor** [reply]
I'm a bit late to the conversation, but I think something to this effect could be nice. 

<img width="1643" alt="Screen Shot 2022-05-31 at 4 46 05 PM" src="https://user-images.githubusercontent.com/47870619/171281731-849c51a2-7113-4f12-b539-ec47283dad26.png">

Replacing the dot with the name of the relation, and placing it above the header. As an added bonus, this would give a larger hit-area for the tooltip.

**Reply by rijkvanzanten** [reply]
cc @benhaynes

**Reply by benhaynes** [reply]
That's a nice solution, @AustinPhillipTaylor! The only issue is the vertical spacing. Sometimes this component is shown within a border, and the gray name extending above the header might be a bit odd. I wouldn't want to shift the main (black) title down, as that would throw off the vertical alignment with other headings.

This seems like the best option beyond the "dot"... but needs a bit more thinking.

**Reply by sharkfin009** [reply]
Was thinking about this , would be great!

**Reply by rasco** [reply]
This looks really good @AustinPhillipTaylor

**Reply by emahuni** [reply]
I hope I am not tooo late for the party. @benhaynes How about this?:

![image](https://github.com/directus/directus/assets/7290508/5ca07678-8c56-4e1b-8475-000eab5e092e)

just add nicer looking icons.

Restore just uses the original field name from the actual db collection.

**Comment by adamkhan** [comment]
[First post from a potential Strapi émigré in wake of their v4 GraphQL changes.]

Any progress with this issue? I on v9.26.0 and have multiple columns labelled "Title"...

**Reply by rijkvanzanten** [reply]
Welcome! 👋🏻 

There's a handful of unanswered questions we have to resolve before we can go over to implementation, like:

- Where are the custom "inline" names stored? On the preset? What if you change the layout you're using?
- How do the translations work for these custom names?
- Should an admin be able to preconfigure these custom inline names somehow?

> [...] in wake of their v4 GraphQL changes.

I'm personally not familiar enough to know what you mean by this. Mind sharing exactly what change you mean (so I can avoid making that change myself? 😅)

**Reply by adamkhan** [reply]
Thanks, @rijkvanzanten!

The GraphQL syntax change in Strapi v4 is a result of deciding to standardize to the JSON:API spec (https://jsonapi.org/format/) with the stated goal of bringing the GraphQL syntax in line with the REST syntax, but this means that every gql query must now be contained within a parent "data" and all fields within a parent "attributes" so that graphql queries are wordier and results deeper. See discussion at https://forum.strapi.io/t/discussion-regarding-the-complex-response-structure-for-rest-graphql-developer-experience.

**Reply by rijkvanzanten** [reply]
Interesting! If such a format is required, I would've expected they'd go with Apollo's style (eg `nodes` / `edges` instead of `data` / `attributes`) to get some compatibility with federation as well, but what do I know 😄

**Comment by MarkBurvs** [comment]
Any progress on this? Would be so handy.

**Comment by allen-wang-rocketlab** [comment]
Just curious if any progress on this. We are really keen to have it.

**Comment by davideghz** [comment]
While waiting for a more complex "rename field" feature, I think a simpler dot notation in the collection header could suffice for the most basic use cases.
![image](https://github.com/user-attachments/assets/eafbc55b-365c-4318-8294-71b442e1f573)

**Comment by kamil-malinski** [comment]
any news on that? some tables are really difficult to use if you have mulitple "name" or "title" columns

**Comment by rajajam** [comment]
Is there any updates on this?

**Comment by poor-john** [comment]
@benhaynes any updates on this?
I know the team have so many big things going on, but simple features like this can also makes a huge impact for us.
I'm even considering developing my own dashboard just because of this (clients would be very confused otherwise).
Thank you for the hard work on Directus, I really love it.

**Comment by achmonth** [comment]
I would really love this as well. For my use case, this is the only feature that keeps me from using Directus.

**Comment by alessioVelluso** [comment]
Does anyone have any updates on this?

**Comment by bencarp** [comment]
Still waiting for this! Seeing how well thought-out Directus usually is, this seems like an unexpected oversight.
